Support GDB JIT API

justinmichaud opened the pull request:
#### bd3a0de8c664246c3bad0429c3b6a5347850682d
<pre>
Support GDB JIT API
<a href="https://bugs.webkit.org/show_bug.cgi?id=291834">https://bugs.webkit.org/show_bug.cgi?id=291834</a>

Reviewed by Yusuke Suzuki.

This adds support for GDB and LLDB to symbolicate backtraces of JIT code. The
dwarf generator  mostly comes from V8, with some modifications to imporve memory safety
and work with WTF. This also adds support for ARM64, ARMv7 and mach-o symbol tables,
which V8 was lacking.

We also add dwarf info to LLInt on ARM64 / ARMv7 to see through LLInt
frames.

This works by using the default __jit_debug_register_code API; A
previous attempt over a decade ago did this using the custom GDB jit
api, which is also probably a valid approach:
<a href="https://bugs.webkit.org/show_bug.cgi?id=75387">https://bugs.webkit.org/show_bug.cgi?id=75387</a>

This approach was chosen because (a) it is what many other jits do, and
(b) because it might be able to be extended in the future to be useful.

We also force frame pointers when DEVELOPER_MODE is enabled; This might
be somewhat controversial, but it is super important for fast backtraces
when profiling too, so I don&apos;t think it is worth the effort to make this
patch support sp-based unwinding.

Canonical link: <a href="https://commits.webkit.org/295194@main">https://commits.webkit.org/295194@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bd3b3a443cb654d741e859402d71d8b1ebe47aee

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104355 "22 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24059 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14474 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109557 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55023 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106395 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24432 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32606 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79237 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107361 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18991 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94173 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59567 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/18789 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12247 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54383 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/97030 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88488 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12301 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111938 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/102966 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31512 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/23216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88277 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31876 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90418 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87942 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22394 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32837 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10601 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25980 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31440 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36765 "Found 2 new failures in jit/GdbJIT.cpp and found 1 fixed file: bytecode/StructureStubClearingWatchpoint.h") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/127169 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31234 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/127169 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34570 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32794 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/assembler/GdbJIT.cpp:
```diff
@@ -0,0 +1,1123 @@
+
+#undef ASSERT_EQ
+
+#endif // ENABLE(ASSEMBLER)
```
Please apply WebKit coding style

Constellation commented on Source/JavaScriptCore/jit/GdbJIT.cpp:
```diff
@@ -0,0 +1,1351 @@
+
+GdbJIT::GdbJIT()
+{
+}
```
```
GdbJIT:: GdbJIT() = default;
```

Constellation commented on Source/JavaScriptCore/jit/GdbJIT.cpp:
```diff
@@ -0,0 +1,1351 @@
+
+template <typename V>
+static inline void WriteUnalignedValue(uint8_t* p, V value) {
+  memcpy(p, &value, sizeof(V));
```
Overall, this file's indent is 2 spaces. That should be 4 spaces.

Constellation commented on Source/JavaScriptCore/jit/GdbJIT.cpp:
```diff
@@ -0,0 +1,1351 @@
+#endif
+
+template <typename V>
+static inline void WriteUnalignedValue(uint8_t* p, V value) {
```
Function name should start with small letter.

Constellation commented on Source/JavaScriptCore/jit/GdbJIT.cpp:
```diff
@@ -0,0 +1,1351 @@
+  };
+
+  template <typename T>
+  void Write(const T& val) {
```
Ditto.

Constellation commented on Source/JavaScriptCore/jit/GdbJIT.cpp:
```diff
@@ -0,0 +1,1351 @@
+
+    T* operator->() { return w_.RawSlotAt<T>(offset_); }
+
+    void set(const T& value) {
```
`{` should be in the next line.

Constellation commented on Source/JavaScriptCore/jit/GdbJIT.cpp:
```diff
@@ -0,0 +1,1351 @@
+// GDB will place breakpoint into this function.
+// To prevent GCC from inlining or removing it we place noinline attribute
+// and inline assembler statement inside.
+static void __attribute__((noinline)) __jit_debug_register_code() { __asm__(""); }
```
Use `NEVER_INLINE`. And we should also use `REFERENCED_FROM_ASM` to make sure that this function will not be removed via LTO etc.
Also, I wonder if we should annotate it with `JS_EXPORT_PRIVATE` to ensure this symbol is visible.
And instead, making it non static.

Constellation replied:
Ah, but maybe, we should not expose it outside of JSC since multiple libraries could have that too. So in the mean time, let's just attach `REFERENCED_FROM_ASM` and `NEVER_INLINE`

Constellation commented on Source/JavaScriptCore/jit/GdbJIT.cpp:
```diff
@@ -0,0 +1,1351 @@
+// GDB will inspect contents of this descriptor.
+// Static initialization is necessary to prevent GDB from seeing
+// uninitialized descriptor.
+JITDescriptor __jit_debug_descriptor = {1, 0, { }, { }};
```
Ditto about `JS_EXPORT_PRIVATE`.

Constellation replied:
Ah, maybe we should avoid exposing it since there are multiple libraries implementing this interface.
I think `REFERENCED_FROM_ASM` would be good to ensure that this is kept even with LTO etc.

Constellation replied:
So, probably,
```
REFERENCED_FROM_ASM JITDescriptor __jit_debug_descriptor = {1, 0, { }, { }};
```

Constellation commented on Source/JavaScriptCore/jit/GdbJIT.cpp:
```diff
@@ -0,0 +1,1351 @@
+  if (align != 0) {
+    for (uint32_t i = 0; i < (sizeof(void*) - align); i++) {
+      w->Write<uint8_t>(DW_CFA_NOP);
+    }
```
We do not need `{` and `}`.

Constellation commented on Source/JavaScriptCore/runtime/OptionsList.h:
```diff
@@ -143,6 +143,7 @@ bool hasCapacityToUseLargeGigacage();
     v(Bool, dumpBBQDisassembly, false, Normal, "dumps disassembly of BBQ wasm code upon compilation"_s) \
     v(Bool, dumpOMGDisassembly, false, Normal, "dumps disassembly of OMG wasm code upon compilation"_s) \
     v(Bool, useJITDump, false, Normal, "generates JITDump side-data") \
+    v(Bool, useGdbJITInfo, false, Restricted, "generates GDB JIT API side-data") \
```
I think `Normal` is fine.

Constellation commented on Source/JavaScriptCore/jit/GdbJIT.cpp:
```diff
@@ -0,0 +1,1351 @@
+0x400848: CFA=WSP
+*/
+
+void UnwindInfoSection::WriteFDEState(Writer* w) {
```
NICE

Constellation reviewed (approved):
Looks good! Awesome

Constellation commented on Source/JavaScriptCore/runtime/Options.cpp:
```diff
@@ -765,6 +765,13 @@ void Options::notifyOptionsChanged()
+        dataLogLn("useGdbJITInfo should be used with OFFLINE_ASM_ALT_ENTRY");
+#endif
+    if (Options::useGdbJITInfo() && Options::useConcurrentJIT())
+        dataLogLn("useGdbJITInfo should be used with JSC_useConcurrentJIT=0");
```
I think this does not make sense as code registration happens before starting using it.

Constellation reviewed (approved):
The other part looks good to me.

mcatanzaro commented on Source/cmake/OptionsCommon.cmake:
```diff
@@ -296,6 +296,10 @@ elseif (NOT ENABLE_ASSERTS)
 
+if (ENABLE_DEVELOPER_MODE)
+    WEBKIT_PREPEND_GLOBAL_COMPILER_FLAGS(-DENABLE_OFFLINE_ASM_ALT_ENTRY=1)
+endif ()
```
What does this actually do? What is the downside to enabling it unconditionally? Performance penalty?

Notably, the stack traces we use to debug bug reports are almost never going to have developer mode enabled. The good news is JavaScriptCore is much more stable than it used to be, so I guess it might not make sense to pay a large performance penalty to get better stack traces. But if the penalty is small, then it's probably worth it.

justinmichaud replied:
I think I will separate this part into a separate PR, and try to measure the results.

mcatanzaro commented on Source/cmake/WebKitCompilerFlags.cmake:
```diff
@@ -136,6 +136,11 @@ if (DEVELOPER_MODE AND DEVELOPER_MODE_FATAL_WARNINGS)
+if (DEVELOPER_MODE AND ARM)
+    # Make sure we get good back traces.
+    WEBKIT_PREPEND_GLOBAL_COMPILER_FLAGS("-fno-omit-frame-pointer")
+endif ()
```
Why only on ARM?

Notably, Fedora also uses `-mno-omit-leaf-frame-pointer` on aarch64 and x86_64, so consider whether you need that too (probably).

justinmichaud replied:
Nice catch!

mcatanzaro reviewed (commented):
I'm quite eager to see how this works.

It's generally not a good idea to enable/disable stuff based on `DEVELOPER_MODE`, though.

justinmichaud committed:
Support GDB JIT API
https://bugs.webkit.org/show_bug.cgi?id=291834

Reviewed by Yusuke Suzuki.

This adds support for GDB and LLDB to symbolicate backtraces of JIT code. The
dwarf generator  mostly comes from V8, with some modifications to imporve memory safety
and work with WTF. This also adds support for ARM64, ARMv7 and mach-o symbol tables,
which V8 was lacking.

We also add dwarf info to LLInt on ARM64 / ARMv7 to see through LLInt
frames.

This works by using the default __jit_debug_register_code API; A
previous attempt over a decade ago did this using the custom GDB jit
api, which is also probably a valid approach:
https://bugs.webkit.org/show_bug.cgi?id=75387

This approach was chosen because (a) it is what many other jits do, and
(b) because it might be able to be extended in the future to be useful.

We also force frame pointers when DEVELOPER_MODE is enabled; This might
be somewhat controversial, but it is super important for fast backtraces
when profiling too, so I don't think it is worth the effort to make this
patch support sp-based unwinding.

Canonical link: https://commits.webkit.org/295194@main
